all: backend separation

div72 opened the pull request.

div72 committed:
all: initial backend separation

furkansimsekli commented on src/bot.py:
```diff
@@ -16,50 +16,41 @@
+        for backend in self.backends.values():
+            backend.run()
+
+        while True:
```
This `while` block might be inefficient since the condition is always `True`. Also, doesn't `time.sleep()` cripple the program?
I'm currently using [APScheduler](https://pypi.org/project/APScheduler/) in a implicit way to schedule a task. python-telegram-bot provides it.

div72 replied:
> This while block might be inefficient since the condition is always True.

What do you mean? The while block lacking an exit condition is a problem but I am not sure if I understand what you mean by inefficient.

> Also, doesn't time.sleep() cripple the program?

The python-telegram-bot version you use(v13) uses threading, so running time.sleep in the main thread won't halt the execution of other threads. That'll be a concern if you upgrade to v20 which uses asyncio but changing the time.sleep to asyncio.sleep would easily fix that.

I thought about using the scheduler but the main bot's implementation must be backend-agnostic therefore we cannot use the telegram backend's scheduler and must initialize our own. And initializing a scheduler just for one task felt a bit wasteful to me but I can do that if you want.

furkansimsekli replied:
>  The while block lacking an exit condition is a problem but I am not sure if I understand what you mean by inefficient.

I guess "inefficiency" isn't the right word here, but people generally don't recommend using it without a proper cause. However, I read some threads on it, now I don't think it's a bad idea.

> I thought about using the scheduler but the main bot's implementation must be backend-agnostic therefore we cannot use the telegram backend's scheduler and must initialize our own. And initializing a scheduler just for one task felt a bit wasteful to me but I can do that if you want.

I see your point on *backend-agnostic* which is nice. Implementing our own scheduler can be helpful if we add new tasks with different intervals. Currently there are only one task but I hope there will be others in the future. See [to-do](https://github.com/furkansimsekli/hu-announcement-bot#to-do).

div72 committed:
all: standardize relative imports

Rationale:
    Running the application from the root requires weird PYTHONPATH
    patches and mypy has a hard time following imports when absolute
    imports are used.

div72 committed:
backend.telegram: fix signature for TelegramUser.send

div72 committed:
all: vendored command parsing/handling

div72 committed:
abc: add user equality

div72 committed:
cmd: use proper names

div72 committed:
cmd_handler: fix admin checks

div72 committed:
abc: add first/last name to user

div72 committed:
cmd_handler: fix enroll

div72 committed:
abc: add subscription methods to User

div72 committed:
backend: delegate message/command handling to Bot

div72 committed:
abc: make a new method to return the bot user

div72 committed:
bot: don't trigger on self-messages

div72 committed:
backend.telegram: properly handle un-registered users

div72 committed:
bot: add event hooks and fix (un)sub commands

div72 committed:
cmd_handler: fix settings cmd

div72 committed:
all: cleanup and fix invocation

div72 committed:
procfile: use python 3 properly

div72 committed:
all: change builtin type annotations to typing

Rationale:
    Using builtin types to type annotate is a Python 3.9+ feature,
    This change makes it so the bot can be used in lower than that.

div72 committed:
cmd_handler: fix typo in start cmd
